DOC: add docstrings for numeric types

YannickJadoul opened the pull request:
This PR adds docstrings for `np.float16`, `np.uint8`, `np.uint16`, `np.uint32`, `np.uint64`. Furthermore, the formatting of the char codes in the existing docstrings is unified to the same format.

Where to generate the autosummary in sphinx, adding it to the reference documentation is not completely clear, though: the current documentation on dtypes is spread out over 2 or 3 different places:
- https://docs.scipy.org/doc/numpy/reference/arrays.scalars.html
- https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html
- https://docs.scipy.org/doc/numpy/user/basics.types.html

See #10106.

(Part of EuroSciPy 2018 sprints)

YannickJadoul committed:
DOC: add docstring for np.float16 (see #10106)

YannickJadoul committed:
DOC: add docstrings for uint8, uint16, uint32, uint64 (see #10106)

YannickJadoul committed:
DOC: make numeric types' docstrings char codes formatting consistent

charris commented:
Test failure is bogus.

jeffyancey commented on numpy/core/_add_newdocs.py:
```diff
@@ -7970,36 +7970,50 @@ def luf(lamdaexpr, *args, **kwargs):
+    """
+    Boolean type (True or False) stored as a byte. Character code: ``'?'``.
+    Alias: bool8.
+
```
not sure it matters, but there is an extra newline here

jeffyancey reviewed (approved):
this looks fine to me.

seberg commented on numpy/core/_add_newdocs.py:
```diff
@@ -8009,22 +8022,65 @@ def luf(lamdaexpr, *args, **kwargs):
 add_newdoc('numpy.core.numerictypes', 'int64',
-    """64-bit integer. Character code 'l'. Python int compatible.""")
+    """
+    64-bit integer. Character code ``'l'``. Python int compatible.
```
Wondering a bit about the Python int compatible part here. It is mostly valid for python2, and even then wrong on windows (and 32bit linux).

eric-wieser replied:
Yeah, this is a) not true on 32-bit windows, and b) doesn't set `__doc__` correctly anyway

mattip replied:
`np.dtype('int64').__doc__` is empty, but `np.int64.__doc__` shows this string.

eric-wieser replied:
My mistake, I was misinterpreting the results of #10106

seberg commented on numpy/core/_add_newdocs.py:
```diff
@@ -8009,22 +8022,65 @@ def luf(lamdaexpr, *args, **kwargs):
     """
-    128-bit floating-point number. Character code: 'g'. C long float
+    128-bit floating-point number. Character code: ``'g'``. C long float
     compatible.
```
c long float (or long double?). Wonder if it should warn here that it is not an IEEE quad float type. And the same text is  true for float96, also... But maybe this is not the place to get into those details. Not sure how prominent this shows up also on the online docs.

seberg reviewed (commented):
Just comments, did not read all, but I think it should all be good even without going into those, just bous questions.

eric-wieser commented:
AFAIR, setting `__doc__` on the types through `add_newdocs` doesn't actually work - but perhaps I'm mistaken.

I think we need to tread carefully here - we might end up wanting different online docs to the local docs - local docs know the exact mapping of sized aliases to C types, whereas the sphinx docs shouldn't make assumptions about the user's machine.

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -8009,22 +8022,65 @@ def luf(lamdaexpr, *args, **kwargs):
 add_newdoc('numpy.core.numerictypes', 'int32',
-    """32-bit integer. Character code 'i'. C int compatible.""")
+    """
+    32-bit integer. Character code ``'i'``. C int compatible.
```
Not true - on my machine, the character code is `l`.

If we want to add these docs, I think we should do it to `byte`/`short`/`intc`/`int_`/`longlong`, not the sized aliases

YannickJadoul commented:
> we might end up wanting different online docs to the local docs - local docs know the exact mapping of sized aliases to C types, whereas the sphinx docs shouldn't make assumptions about the user's machine.

OK, this is an important point for good documentation. I will take a second stab, keeping this in mind.

To get a better picture: are these docstrings meant to also replace [the list of scalar dtypes in the docs](https://docs.scipy.org/doc/numpy/reference/arrays.scalars.html) on the long run (i.e., with `autosummary` generated documentation) ? What level of detail is needed, and in what context should these docstrings present in the online documentation?

eric-wieser commented:
Perhaps the correct path for now is:

* Document the native types using `add_new_docs`, with something like
  > Corresponds to the C `long` type
* Add a generated sentence like
  > **On the machine used to generate these docs**, this type is aliased to `np.int64` <docs for int64>

YannickJadoul committed:
DOC: Adding documentation to C/Python compatible names of scalar types

YannickJadoul committed:
DOC: Adding platform-dependent aliases to scalar type docstrings

YannickJadoul commented:
I'm not completely happy with how this sketch of a solution looks, but it should be a step closer to the actual correct, yet platform-flexible documentation. Is this where this PR wants to go?

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7969,62 +7969,199 @@ def luf(lamdaexpr, *args, **kwargs):
+
+integer_aliases = type_aliases('numpy.core.numerictypes', [
+    ('int8', '8-bit singed integer (-128 to 127)'),
+    ('int16', '16-bit singed integer (-32768 to 32767)'),
```
typo: signed

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7969,62 +7969,199 @@ def luf(lamdaexpr, *args, **kwargs):
+    ('uint16', '16-bit unsinged integer (0 to 65535)'),
+    ('uint32', '32-bit unsinged integer (0 to 4294967295)'),
+    ('uint64', '64-bit unsinged integer (0 to 18446744073709551615)'),
+    ('uintp', 'Unsigned integer large enough to fit pointer, compatible with C ``size_t``'),
```
Strictly it's `uintptr_t`

YannickJadoul replied:
Changed `size_t` and `ssize_t` to `intptr_t` and `uintptr_t`, but the reference docs are again slightly wrong, here, then: https://docs.scipy.org/doc/numpy/user/basics.types.html

Should these be fixed as well?

eric-wieser commented:
I like the direction this is going in - I think it will definitely be an improvement over what already exists.

I think we'll need to think a little more about how to use this with server-built sphinx builds - but for local users using `help(type)`, I think this does exactly the right thing.

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7969,62 +7969,199 @@ def luf(lamdaexpr, *args, **kwargs):
+    {}
+    """.format(extra_alias_doc('numpy.core.numerictypes', 'single', float_aliases)))
 
+add_newdoc('numpy.core.numerictypes', 'float_',
```
Strictly this should be `double` - `float_` is an alias defined to be compatible with python `float`, although admittedly the two are always the same.

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7969,62 +7969,199 @@ def luf(lamdaexpr, *args, **kwargs):
+    {}
+    """.format(extra_alias_doc('numpy.core.numerictypes', 'float_', float_aliases)))
+
+add_newdoc('numpy.core.numerictypes', 'longfloat',
```
I think `longdouble` is the canonical name.

YannickJadoul replied:
OK, just followed the list here: https://docs.scipy.org/doc/numpy/reference/arrays.scalars.html. But indeed, `longfloat` seems to be an alias for `longdouble`.

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7969,62 +7969,199 @@ def luf(lamdaexpr, *args, **kwargs):
+    {}
+    """.format(extra_alias_doc('numpy.core.numerictypes', 'clongfloat', complex_aliases)))
+
+add_newdoc('numpy.core.numerictypes', 'complex192',
```
This one is an alias, not a real type

YannickJadoul replied:
Does that mean that `float96`is also an alias, and either `float96` or `float128` is matched against `longdouble`?

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7969,62 +7969,199 @@ def luf(lamdaexpr, *args, **kwargs):
+def type_aliases(place, aliases):
+    return [(alias_type, doc) for (alias, doc) in aliases
+                for alias_type in (get_type(place, alias),)
+                if alias_type is not None]
```
Rather than using a hack to make a local variable in a comprehension, I think this would be clearer as a generator using `yield`

YannickJadoul committed:
DOC: Fix typos and minor issues in scalar type docstrings

YannickJadoul commented:
Update: fixed minor things, but:

- `clongfloat` is mentioned in the existing documentation, instead of `clongdouble`.
- The reference documentation on `intp` mentions `ssize_t` instead of `intptr_t`.

But these can probably be fixed once these docstrings get included in the reference documentation?

Next, stackoverflow shows a few hackish way of detecting sphinx (https://stackoverflow.com/questions/20843737/check-if-sphinx-doc-called-the-script), but something feels wrong about actually accessing and using this when generating the docs?

eric-wieser commented:
Lets leave worrying about sphinx to a later PR.

Both your points are correct - but I think the other docs are just subtly wrong, and would prefer to aim for correctness not consistency.

eric-wieser reviewed (commented):
Can you include both the name and doc in the list of aliases? Right now I think you only add the doc.

Also, adding `float_` and `complex_` to the list of aliases would be handy.

Can you show the output of `help(np.intc)`, `help(np.long)`, and `help(np.longlong)` after this change?

I suspect there's some overlap here with #10151

YannickJadoul commented:
> Can you include both the name and doc in the list of aliases? Right now I think you only add the doc.

Yes, forgot that, thanks!

> Also, adding float_ and complex_ to the list of aliases would be handy

Since these seem to be ['hardcoded' aliases](https://github.com/numpy/numpy/blob/master/numpy/core/numerictypes.py#L406), I'm adding these as part of the docstring itself, and not with this alias list construct.

YannickJadoul committed:
DOC: Add hard-coded aliases to scalar type docstrings

YannickJadoul commented:
> Can you show the output of help(np.intc), help(np.long), and help(np.longlong) after this change?

```PyCon
>>> print(np.intc.__doc__)
Signed integer type, compatible with C ``int``.
    Character code: ``'i'``.
>>> print(np.long.__doc__)
int(x=0) -> integer
int(x, base=10) -> integer

Convert a number or string to an integer, or return 0 if no arguments
are given.  If x is a number, return x.__int__().  For floating point
numbers, this truncates towards zero.

If x is not a number or if base is given, then x must be a string,
bytes, or bytearray instance representing an integer literal in the
given base.  The literal can be preceded by '+' or '-' and be surrounded
by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
Base 0 means to interpret the base from the string as an integer literal.
>>> int('0b100', base=0)
4
>>> print(np.longlong.__doc__)
Signed integer type, compatible with C ``long long``. 
    Character code: ``'q'``.
```

I'll have a closer look at that other PR later.

eric-wieser commented:
I'd specifically like to see the output of `help`. Also I made a typo, and said `np.long` not `np.int_`.

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7969,62 +7969,198 @@ def luf(lamdaexpr, *args, **kwargs):
+    try:
+        return getattr(__import__(place, globals(), {}, [obj]), obj)
+    except Exception:
+        return None
```
I think this should be `AttributeError`.

YannickJadoul replied:
Stolen from `add_newdoc` in `core/function_base.py`, where it also says `Exception`, but I've adapted it.

YannickJadoul commented:
> I'd specifically like to see the output of help.

Sorry, here you go:

> `help(np.intc)`

```
Help on class int32 in module numpy:

class int32(signedinteger)
 |  Signed integer type, compatible with C ``int``.
 |  Character code: ``'i'``.
 |  
 |  Method resolution order:
 |      int32
 |      signedinteger
 |      integer
 |      number
 |      generic
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
[...]
 |  astype(...)
 |      Not implemented (virtual attribute)
 |      
 |      Class generic exists solely to derive numpy scalars from, and possesses,
 |      albeit unimplemented, all the attributes of the ndarray class
 |      so as to provide a uniform API.
 |      
 |      See Also
 |      --------
 |      The corresponding attribute of the derived class of interest.
 |  
[...]
```

> `help(np.int_)

```
Help on class int64 in module numpy:

class int64(signedinteger)
 |  Signed integer type, compatible with Python `int` anc C ``long``.
 |  Character code: ``'l'``.
 |  
 |  Method resolution order:
 |      int64
 |      signedinteger
 |      integer
 |      number
 |      generic
 |      builtins.object
[...]
```

> `help(np.longlong)`

```
Help on class int64 in module numpy:

class int64(signedinteger)
 |  Signed integer type, compatible with C ``long long``. 
 |  Character code: ``'q'``.
 |  
 |  Method resolution order:
 |      int64
 |      signedinteger
 |      integer
 |      number
 |      generic
 |      builtins.object
[...]
```

eric-wieser commented:
It looks like your alias interpolation is not working then - I don't see any sign of it in that output.

YannickJadoul committed:
DOC: Fix platform-dependent aliases in doc string of scalar types

YannickJadoul commented:
> It looks like your alias interpolation is not working then - I don't see any sign of it in that output.

*Sigh* ... yeah, changing working code to compare the alias string and the actual type is ... not working of course.

I'll add a unit test tomorrow, but I need to figure out where these are and how they're organized, first.

Help now looks like:
```
class int32(signedinteger)
 |  Signed integer type, compatible with C ``int``.
 |  Character code: ``'i'``.
 |  Alias *on this platform*: int32: 32-bit signed integer (-2147483648 to 2147483647).
```

```
class int64(signedinteger)
 |  Signed integer type, compatible with Python `int` anc C ``long``.
 |      Character code: ``'l'``.
 |      Alias *on this platform*: int64: 64-bit signed integer (-9223372036854775808 to 9223372036854775807).
 |  Alias *on this platform*: intp: Signed integer large enough to fit pointer, compatible with C ``intptr_t``. Alias: int0.
```

```
Help on class int64 in module numpy:

class int64(signedinteger)
 |  Signed integer type, compatible with C ``long long``. 
 |  Character code: ``'q'``.
```

So something weird's going on with the indentation as well. So I'll look at that as well, tomorrow, when finishing up things.

YannickJadoul committed:
DOC: Fix indentation of numeric types' docstring

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7969,62 +7969,198 @@ def luf(lamdaexpr, *args, **kwargs):
+    Signed integer type, compatible with Python `int` anc C ``long``.
+    Character code: ``'l'``.
+    {}
+    """.format(extra_alias_doc('numpy.core.numerictypes', 'int_', integer_aliases)))
```
Wrapping the format string in `textwrap.dedent` would solve your indentation problem

YannickJadoul replied:
Apparently indenting everything equally also does the job (guessing Python's help does the `dedent`, then), see fresh commit `8a53e9d`. (And that's how it already was before this PR as well.)

eric-wieser replied:
`add_newdocs` does a dedent internally

eric-wieser commented:
I think the test could go in `numpy\core\tests\test_numerictypes.py`

YannickJadoul committed:
TST: Add test checking for platform-dependent generation of numeric type docstrings

YannickJadoul committed:
TST: Fix docstring test failures when PYTHONOPTIMIZE/Py_OptimizeFlag > 1

charris commented:
Looks like this needs a rebase.

eric-wieser commented:
@YannickJadoul: Looks like you just clobbered the small fixup I made, so I'll add a comment inline instead

eric-wieser commented on numpy/core/tests/test_numerictypes.py:
```diff
@@ -444,3 +444,14 @@ def test_complex(self, t):
+    def test_platform_dependent_aliases(self):
+        if sys.flags.optimize > 1:
+            # Docstrings do not get added if PYTHONOPTIMIZE/Py_OptimizeFlag > 1
+            return
```
Use `@pytest.mark.skipif(sys.flags.optimize > 1, "no docstrings present to inspect")` or similar here

eric-wieser commented:
Can you show the output of the three `help`s once more?

YannickJadoul commented:
> @charris: Conflict was my fault (and trivial to resolve) - resolved using the web UI
> @YannickJadoul: Looks like you just clobbered the small fixup I made, so I'll add a comment inline instead

Oh, shit, sorry. Bad timing and too happily force-pushing, there. I'll quickly add that changes.

I've also taken a look at #10151, but if that PR is only changing the names of the type objects, there should be no direct conflicts with this PR. Except of course if these types and their documentation are to be further restructured.

eric-wieser commented:
> Bad timing and too happily force-pushing, there.

Don't worry about it

> and too happily force-pushing, there.

For future reference, `--force-with-lease` is the solution to this kind of problem - it will let you push over changes you know exist, but not ones you don't

YannickJadoul committed:
TST: Moving check of PYTHONOPTIMIZE for test_numerictypes.py::TestDocStrings into pytest.mark.skip decorator

eric-wieser commented:
#10151 doesn't block this issue.

One thing that would be nice: that right now, `help(np.ushort)` does not actually contain the text `np.ushort` anywhere!

YannickJadoul commented:
> that right now, help(np.ushort) does not actually contain the text np.ushort anywhere!

The one thing we could do is add that also as an 'alias' or 'name' in the docstring? This would probably have the advantage of also working in the opposite direction if someone asks e.g. for `help(np.int32)`.

YannickJadoul commented:
> `help(np.intc)`

```
Help on class int32 in module numpy:

class int32(signedinteger)
 |  Signed integer type, compatible with C ``int``.
 |  Character code: ``'i'``.
 |  Alias *on this platform*: int32: 32-bit signed integer (-2147483648 to 2147483647).
```

> `help(np.int_)`

```
Help on class int64 in module numpy:

class int64(signedinteger)
 |  Signed integer type, compatible with Python `int` anc C ``long``.
 |  Character code: ``'l'``.
 |  Alias *on this platform*: int64: 64-bit signed integer (-9223372036854775808 to 9223372036854775807).
 |  Alias *on this platform*: intp: Signed integer large enough to fit pointer, compatible with C ``intptr_t``. Alias: int0.
```

> `help(np.longlong)`

```
Help on class int64 in module numpy:

class int64(signedinteger)
 |  Signed integer type, compatible with C ``long long``. 
 |  Character code: ``'q'``.
```

> that right now, help(np.ushort) does not actually contain the text np.ushort anywhere!

Actually, I've just realized: it's even worse/weirder than that. `help(np.intc)` says (on my computer) that it's `np.int32` (OK there, I think), but then says that the type with name `int32` has _on this platform_ an alias called `np.int32`!?

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7970,62 +7970,198 @@ def luf(lamdaexpr, *args, **kwargs):
+    ('uint16', '16-bit unsigned integer (0 to 65535)'),
+    ('uint32', '32-bit unsigned integer (0 to 4294967295)'),
+    ('uint64', '64-bit unsigned integer (0 to 18446744073709551615)'),
+    ('uintp', 'Unsigned integer large enough to fit pointer, compatible with C ``uintptr_t``. Alias: uint0'),
```
I don't think we should document `int0` and `uint0` - I've never seen any rationale behind their names (#11025), and I'd prefer that no-one ever use them.

eric-wieser commented:
> Actually, I've just realized: it's even worse/weirder than that. `help(np.intc)` says (on my computer) that it's `np.int32` (OK there, I think), but then says that the type with name `int32` has _on this platform_ an alias called `np.int32`!?

Yeah, it's a mess. You'll notice that in your above output, `long_` and `longlong` both claim to be showing the docs for `np.int64`, yet show different docs.

How about setting the first line to "The docs below describe the `intc` type, not the `int32` alias for this type", and similarly for all the other types.

YannickJadoul committed:
DOC: Cleanup code for adding docstrings to numeric types and mention canonical type name

YannickJadoul commented:
> How about setting the first line to "The docs below describe the `intc` type, not the `int32` alias for this type", and similarly for all the other types.

Myeah, that should be possible, but I'm not completely convinced of how to say it. This feels very intrusive and a little bit too 'meta' as a first line?

I've just cleaned up a little bit and added a line for the "Canonical name" of the type, whenever that differs from `__name__`, as a first step. (See diff between the docstrings in `np.core.numerictypes`, before and after the last commit: [old_new.diff.txt](https://github.com/numpy/numpy/files/2385896/old_new.diff.txt))

eric-wieser commented:
Looks pretty good to me now. Two minor requests:

* It looks like the indent is messed up in a few places in that diff
* Can you prepend `np.` to all of your alias names?

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7970,62 +7970,169 @@ def luf(lamdaexpr, *args, **kwargs):
+    {doc}
+    Character code: ``'{character_code}'``.
+    {canonical_name_doc}{alias_doc}
+    """.format(doc=doc.strip(), character_code=character_code,
```
You'll find the indentation a lot easier to manage if you use textwrap dedent here - in general you want to dedent before doing any formatting

YannickJadoul replied:
I'm a bit weary to dedent everything, because other docstrings seem to be indented as well?
E.g.: 

```pycon
>>> pprint.pprint(np.dtype.__doc__)
('dtype(obj, align=False, copy=False)\n'
 '\n'
 '    Create a data type object.\n'
 '\n'
 '    A numpy array is homogeneous, and contains elements described by a\n'
 '    dtype object. A dtype object can be constructed from different\n'
 '    combinations of fundamental numeric types.\n'
 '\n'
 '    Parameters\n'
 '    ----------\n'
 '    obj\n'
 '        Object to be converted to a data type object.\n'
[...]
```

eric-wieser replied:
I wasn't aware of that, interesting find. I'd assumed that we dedented them in addnewdocs, but it seems that python does it automatically in `help`. I think ultimately the indentation doesn't matter, so what you have is fine.

YannickJadoul replied:
The whole pasting-strings-together is a bit of a mess; you're right there, but until there's more changes that need to be done, this might be the easiest

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7970,62 +7970,169 @@ def luf(lamdaexpr, *args, **kwargs):
+    ])
+
+
+def add_newdoc_for_numeric_type(obj, character_code, fixed_aliases, possible_aliases, doc):
```
Instead of passing the character code, you can just use `np.dtype(o).char` below

YannickJadoul replied:
Same here, actually. To do that, I have to import numpy while numpy is being imported? :-/

eric-wieser replied:
Numpy is already full of import inter-dependencies. I don't have a simple description of the rules, but as long as there are no cycles at import time, things are usually ok

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7970,62 +7970,169 @@ def luf(lamdaexpr, *args, **kwargs):
 
-add_newdoc('numpy.core.numerictypes', 'bool_',
-    """NumPy's Boolean type.  Character code: ``?``.  Alias: bool8""")
+def get_type(place, obj):
```
This `place` argument seems kind of redundant. Perhaps just add `import numpy.core.numerictypes as _types` outside these functions, and then change this function to `getattr(_types, obj, None)`

YannickJadoul replied:
Looks good, but the import fails when I do that:
```
>>> import numpy as np
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/yannick/numpy/env/lib/python3.6/site-packages/numpy-1.16.0.dev0+2c80d46-py3.6-linux-x86_64.egg/numpy/__init__.py", line 142, in <module>
    from . import core
  File "/home/yannick/numpy/env/lib/python3.6/site-packages/numpy-1.16.0.dev0+2c80d46-py3.6-linux-x86_64.egg/numpy/core/__init__.py", line 64, in <module>
    from . import _add_newdocs
  File "/home/yannick/numpy/env/lib/python3.6/site-packages/numpy-1.16.0.dev0+2c80d46-py3.6-linux-x86_64.egg/numpy/core/_add_newdocs.py", line 7973, in <module>
    import numpy.core.numerictypes as _types
AttributeError: module 'numpy' has no attribute 'core' 
```

YannickJadoul replied:
The trick's stolen from `add_newdoc` in `numpy/core/function_base.py`, btw: https://github.com/numpy/numpy/blob/master/numpy/core/function_base.py#L366 It's semi-hocuspocus, to me, but I'm assuming that since the `_add_newdocs.py` is run while importing, the submodule is not completely there yet to be imported in the normal way?

YannickJadoul replied:
`from numpy.core import numerictypes as _types` does work, so I can use that.

YannickJadoul committed:
DOC: Cleaning up access of numeric type objects and character codes in dynamic docstring generation

YannickJadoul committed:
DOC: Prepend 'np.' to numeric type docstrings

YannickJadoul commented:
> - It looks like the indent is messed up in a few places in that diff
> - Can you prepend np. to all of your alias names in the docstring?

Thanks! Both  should be fixed again, now:

```pycon
>>> help(np.intc)
Help on class int32 in module numpy:

class int32(signedinteger)
 |  Signed integer type, compatible with C ``int``.
 |  Character code: ``'i'``.
 |  Canonical name: ``np.intc``.
 |  Alias *on this platform*: ``np.int32``: 32-bit signed integer (-2147483648 to 2147483647).
[...]
```

eric-wieser reviewed (commented):
Can't comment inline on mobile: Line 8017 should not be passing a default to getattr - the attribute will always exist for canonical names

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7970,62 +7972,164 @@ def luf(lamdaexpr, *args, **kwargs):
+def add_newdoc_for_numeric_type(obj, fixed_aliases, possible_aliases, doc):
+    o = getattr(_numerictypes, obj, None)
+    if o is None:
+        return
```
From my earlier comment - change this to `o = getattr(_numerictypes, obj)`, since the object is guaranteed to exist.

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7970,62 +7972,164 @@ def luf(lamdaexpr, *args, **kwargs):
+        for (alias, doc) in aliases:
+            alias_type = getattr(_numerictypes, alias, None)
+            if alias_type is not None:
+                yield (alias_type, alias, doc)
```
Strictly speaking this would be better as:
```
try:
    alias_type = getattr(_numerictypes, alias)
except AttributeError:
    pass
else:
    yield (alias_type, alias, doc)
```

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7970,62 +7972,164 @@ def luf(lamdaexpr, *args, **kwargs):
 
-add_newdoc('numpy.core.numerictypes', 'object_',
-    """Any Python object.  Character code: 'O'.""")
+add_newdoc_for_numeric_type('object_', [], [],
```
Not sure I'd consider this numeric. Probably better either to rename the function to `*_scalar_type`, or leave the non-numeric types as they were.

eric-wieser committed:
MAINT: Remove dead code, rename function

eric-wieser reviewed (approved):
Made the changes I suggested below myself. Feel free to tweak if you want, but I think this is ready to go in

YannickJadoul committed:
DOC: Cleaning up numeric_type_aliases in core/_add_newdocs.py

YannickJadoul commented:
> Made the changes I suggested below myself. Feel free to tweak if you want, but I think this is ready to go in

I've still implemented the one small change in `numeric_type_aliases` you suggested.

YannickJadoul commented:
Oh, and almost forgot, but I yesterday noticed that this actually doesn't work for the complex types. These types already have a docstring set in the C code, and `add_newdocs` will not overwrite the existing one (but does so silently):

```pycon
>>> help(np.cdouble)
Help on class complex128 in module numpy:

class complex128(complexfloating, builtins.complex)
 |  Composed of two 64 bit floats
```

 I guess I've missed this since been focusing on the signed and unsigned integers, since they had these C type equivalence issues.

The line with this docstring is here: https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/scalartypes.c.src#L3773. My instinctive reaction would be to make this consistent and remove the docstring from the C code, but then I don't know the reason why these docstrings are there?

YannickJadoul committed:
DOC: Fixing dynamic aliases in docstrings for complex scalar types

eric-wieser commented:
> My instinctive reaction would be to make this consistent and remove the docstring from the C code

Seems sensible to me. Even if there is a reason for them to be there, it would apply to all of the docstrings anyway, not just that one,

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7970,62 +7972,165 @@ def luf(lamdaexpr, *args, **kwargs):
+    ('complex128', 'Complex number type composed of 2 64-bit-precision floating-point numbers'),
+    ('complex192', 'Complex number type composed of 2 96-bit extended-precision floating-point numbers'),
+    ('complex256', 'Complex number type composed of 2 128-bit extended-precision floating-point numbers'),
+    ])
```
Is there any real value to having these four separate lists of aliases, rather than building one big lookup?

YannickJadoul replied:
Euhm, yeah, the idea was not not loop over irrelevant aliases (while. But that does seem to come at the cost of being slightly more error-prone. I'm guessing you prefer the side of the trade-off that's less error-prone?

eric-wieser replied:
Yeah, especially since collectivey we've proven that that type of error is easy to make and hard to spot. Thanks!

YannickJadoul replied:
One less argument to `add_newdoc_for_scalar_type` as well, as bonus.

YannickJadoul committed:
DOC: Removing docstrings for complex scalar types in numpy/core/src/multiarray/scalartypes.c.src, as they are now set in numpy/core/_add_newdocs.py

YannickJadoul committed:
DOC: Merging lists of different dynamic aliases into one, and prepending 'np.' to fixed aliases

eric-wieser commented on numpy/core/_add_newdocs.py:
```diff
@@ -7970,62 +7972,156 @@ def luf(lamdaexpr, *args, **kwargs):
+            try:
+                alias_type = getattr(_numerictypes, alias)
+            except AttributeError:
+                pass
```
Could do with a comment here like "the set of aliases that actually exist varies between platforms"

eric-wieser reviewed (approved):
If you want more visibility for this, a release note under "improvements" mentioning that `help(np.intp)` or similar now shows a list of common type aliases would seem pretty sensible

YannickJadoul committed:
DOC: Adding scalar type docstring improvements to release notes

YannickJadoul commented:
Almost done! :-)

Running the code from #10106 again, there are a few more undocumented numeric types that seem to be related, part of the type hierarchy of scalar types:
* [ ] `<class 'numpy.complexfloating'>` (`np.complexfloating`)
* [ ] `<class 'numpy.flexible'>` (`np.flexible`)
* [ ] `<class 'numpy.floating'>` (`np.floating`)
* [ ] `<class 'numpy.inexact'>` (`np.inexact`)
* [ ] `<class 'numpy.integer'>` (`np.integer`)
* [ ] `<class 'numpy.number'>` (`np.number`)
* [ ] `<class 'numpy.signedinteger'>` (`np.signedinteger`)
* [ ] `<class 'numpy.unsignedinteger'>` (`np.unsignedinteger`)

Futhermore, `numpy.void` seems like it could still be a couple of lines included in this PR as well, just after `object` ?

mattip commented:
Perhaps let this go in as-is, and add the rest in another PR?

eric-wieser commented:
I'm with @mattip on this one - documenting the abstract types seems like a separate task to documenting the concrete ones.

eric-wieser committed:
DOC: Tweak release notes

charris commented:
Thanks @YannickJadoul. And thanks to Eric for helping get this knocked into shape.

YannickJadoul commented:
Thanks indeed, @eric-wieser, for the guidance and remarks!
